components/ws-daemon/pkg/cpulimit/dispatch.go

```diff
@@ -214,3 +220,18 @@ func (d *DispatchListener) WorkspaceUpdated(ctx context.Context, ws *dispatch.Wo
 
 	return nil
 }
+
+func newCFSController(basePath, cgroupPath string) (CFSController, error) {
+	controllers := filepath.Join(basePath, "cgroup.controllers")
+	_, err := os.Stat(controllers)
+
+	if os.IsNotExist(err) {
+		return CgroupV1CFSController(filepath.Join(basePath, "cpu", cgroupPath)), nil
+	}
+
+	if err == nil {
+		return CgroupV2CFSController(filepath.Join(basePath, cgroupPath)), nil
+	}
+
+	return nil, err
+}
```

Comment:
@csweichel @Furisto 
It is not very good to have scattered cgroup v2 decisions, so why not unify them since we will be using the runc library? k8s seems to use the same determination.
https://github.com/kubernetes/kubernetes/blob/6eca8cc472ee140de8fc7181a4349df643a86e04/pkg/kubelet/cm/cgroup_manager_linux.go#L152
```suggestion
func newCFSController(basePath, cgroupPath string) (CFSController, error) {
	if cgroups.IsCgroup2UnifiedMode() {
		return CgroupV2CFSController(filepath.Join(basePath, cgroupPath)), nil
	} else {
		return CgroupV1CFSController(filepath.Join(basePath, "cpu", cgroupPath)), nil
	}
}
```

Reply:
Also, since v1 and v2 can coexist, you need to make sure that cpu is enabled in `cgroup.controllers`.

Reply:
Personally I found the runc/containerd cgroup library very hard to use. It's just so much easier for many operations to implement them on a filesystem level. rootlesskit came to the same conclusion.

Reply:
> Also, since v1 and v2 can coexist, you need to make sure that cpu is enabled in `cgroup.controllers`.

They can coexist, but if v1 has the cpu controller attached, then we would need to unmount the v1 cgroup mount in order to make the cpu controller available for v2 which would only work if we somehow managed to remove all processes from the cgroup v1 hierarchy. It would also be a pretty unusual custom setup. The usual options are either pure v1, pure v2 or hybrid mode which is basically v1 (all controllers are attached to v1) with some features available for v2 (this would be detected as v1 in this implementation which is correct because that is where the controllers are).

I agree though that it could theoretically be possible that we have cgroup v2 setup where the cpu controller has not been activated. I will handle that case.

Reply:
I'd argue we settle for pure v1 or pure v2. All other modes result in undefined behaviour.

Reply:
We are currently running hybrid mode on our own nodes. This is the default for Ubuntu. It is not a big deal though. As I said, for all intents and purposes hybrid mode is v1.

Reply:
> We are currently running hybrid mode on our own nodes.

We ran V1 in the nodes (`systemd.unified_cgroup_hierarchy=0`). At the time, v2 broke the CPU limiting

Reply:
I think this is the best.
> I'd argue we settle for pure v1 or pure v2. All other modes result in undefined behaviour.

Reply:
@Furisto @csweichel 
So that is the situation with rootless. I did not know that. I would like to take advantage of that lesson. My root concern is not the using the runc and contained libraries, but the implementation of the determination of whether or not it is cgroup v2 at every point in the future. How about we take a lesson from rootless and have a cgroup v2 check to implement original our codes somewhere(e.g. utils.go or something)?

> Personally I found the runc/containerd cgroup library very hard to use. It's just so much easier for many operations to implement them on a filesystem level. rootlesskit came to the same conclusion.

Reply:
Indeed it makes sense to group the cgroupv2 decision in one place. We could introduce a `cgroups` package and put the check in there. Alternatively, we could move it to the `cpulimit` package, as that one's a leaf in the dependency tree.